pkg/controller/registry/resolver/predicates.go

```diff
@@ -0,0 +1,376 @@
+type gvkPredicate struct {
+	api opregistry.APIKey
+}
+
+func ProvidingAPIPredicate(api opregistry.APIKey) OperatorPredicate {
+	return gvkPredicate{
+		api: api,
+	}
+}
+
+func (g gvkPredicate) Test(o *Operator) bool {
+	for _, p := range o.Properties() {
+		if p.Type != opregistry.GVKType {
+			continue
+		}
+		var prop opregistry.GVKProperty
+		err := json.Unmarshal([]byte(p.Value), &prop)
+		if err != nil {
+			continue
+		}
+		if prop.Kind == g.api.Kind && prop.Version == g.api.Version && prop.Group == g.api.Group {
+			return true
+		}
+	}
+	return false
+}
+
+func (g gvkPredicate) String() string {
+	return fmt.Sprintf("with group: %s, version: %s, kind: %s", g.api.Group, g.api.Version, g.api.Kind)
+}
+
+type skipRangeIncludesPredication struct {
+	version semver.Version
+}
+
+func SkipRangeIncludesPredicate(version semver.Version) OperatorPredicate {
+	return skipRangeIncludesPredication{version: version}
+}
+
+func (s skipRangeIncludesPredication) Test(o *Operator) bool {
+	semverRange, err := o.SemverRange()
+	return err == nil && semverRange(s.version)
+}
+
+func (s skipRangeIncludesPredication) String() string {
+	return fmt.Sprintf("skip range includes: %v", s.version.String())
+}
+
+type replacesPredicate string
+
+func ReplacesPredicate(replaces string) OperatorPredicate {
+	return replacesPredicate(replaces)
+}
+
+func (r replacesPredicate) Test(o *Operator) bool {
+	if o.Replaces() == string(r) {
+		return true
+	}
+	for _, s := range o.Skips() {
+		if s == string(r) {
+			return true
+		}
+	}
+	return false
+}
+
+func (r replacesPredicate) String() string {
+	return fmt.Sprintf("replaces: %v", string(r))
+}
+
+type andPredicate struct {
+	predicates []OperatorPredicate
+}
+
+func And(p ...OperatorPredicate) OperatorPredicate {
+	return andPredicate{
+		predicates: p,
+	}
+}
+
+func (p andPredicate) Test(o *Operator) bool {
+	for _, predicate := range p.predicates {
+		if predicate.Test(o) == false {
+			return false
+		}
+	}
+	return true
+}
+
+func (p andPredicate) String() string {
+	var b bytes.Buffer
+	for i, predicate := range p.predicates {
+		b.WriteString(predicate.String())
+		if i != len(p.predicates)-1 {
+			b.WriteString(" and ")
+		}
+	}
+	return b.String()
+}
+
+func Or(p ...OperatorPredicate) OperatorPredicate {
+	return orPredicate{
+		predicates: p,
+	}
+}
+
+type orPredicate struct {
+	predicates []OperatorPredicate
+}
+
+func (p orPredicate) Test(o *Operator) bool {
+	for _, predicate := range p.predicates {
+		if predicate.Test(o) == true {
+			return true
+		}
+	}
+	return false
+}
+
+func (p orPredicate) String() string {
+	var b bytes.Buffer
+	for i, predicate := range p.predicates {
+		b.WriteString(predicate.String())
+		if i != len(p.predicates)-1 {
+			b.WriteString(" or ")
+		}
+	}
+	return b.String()
+}
+
+type OperatorPredicateFunc func(*Operator) bool
+
+func (opf OperatorPredicateFunc) Test(o *Operator) bool {
+	return opf(o)
+}
+
+func (opf OperatorPredicateFunc) String() string {
+	return ""
+}
```

Comment:
Now that predicates must provide a human-readable string representation, this type can probably be removed altogether. That avoids the risk down the line of a user seeing an empty string and becoming confused.

Reply:
@benluddy I'd explored removing this type for a `genericPredicate` instead, but it seemed to do the same thing, with it's string representation being `""`. The Generic Predicate would be required to implement predicates like the [True](https://github.com/operator-framework/operator-lifecycle-manager/pull/2258/files#diff-a5735698ad5d9d46d8568de6aaddf740642e9c19598888768c68db7d1eec1fe1R298), [False](https://github.com/operator-framework/operator-lifecycle-manager/pull/2258/files#diff-a5735698ad5d9d46d8568de6aaddf740642e9c19598888768c68db7d1eec1fe1R304) predicate etc